sbi/inference/snpe/snpe_base.py

```diff
@@ -420,37 +409,19 @@ def _train(
 
             self._val_log_prob = log_prob_sum / num_validation_examples
 
-            if self._show_progressbar:
-                # end="\r" deletes the print statement when a new one appears.
-                # https://stackoverflow.com/questions/3419984/
-                print("Training neural network. Epochs trained: ", epoch, end="\r")
-
-            if epoch >= max_num_epochs:
-                break
-
-        if self._show_progressbar and self._has_converged(epoch, stop_after_epochs):
-            # network has converged, we print this summary.
-            print("Neural network successfully converged after", epoch, "epochs.")
-        elif self._show_progressbar and max_num_epochs == epoch:
-            # training has stopped because of max_num_epochs argument.
-            print("Stopping neural network training after", epoch, "epochs.")
-
-        if max_num_epochs == epoch:
-            # warn if training was not stopped by early stopping
-            warnings.warn(
-                "Maximum number of epochs reached, but network has not yet "
-                "fully converged. Consider increasing the value of max_num_epochs."
-            )
+            self._maybe_show_progress(self._show_progressbar, epoch)
+
+        self._report_convergence_at_end(epoch, stop_after_epochs, max_num_epochs)
 
         # Update summary.
         self._summary["epochs"].append(epoch)
         self._summary["best_validation_log_probs"].append(self._best_val_log_prob)
 
-    def _loss(self, round_idx: int, theta: Tensor, x: Tensor, masks: Tensor) -> Tensor:
-        """Return the right calibrated loss, depending on the round.
+    def _loss(self, round_idx: int, theta: Tensor, x: Tensor, masks: Tensor,) -> Tensor:
+        """Return a loss calibrated correctly for the first and for further rounds.
```

Comment:
I do not like using the word 'calibrated' here. It just reminds me of the calibration kernel, but this is not what is meant here. I suggest to simply say:

`Return the loss for the first and further rounds.`